File: client/android/div/src/main/java/com/yandex/div/core/util/mask/CurrencyInputMask.kt
package com.yandex.div.core.util.mask

import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.text.NumberFormat
import java.util.Locale
import kotlin.math.abs

internal class CurrencyInputMask(
    locale: Locale,
    private val onError: ((Exception) -> Unit)
) : BaseInputMask(MaskData("", emptyList(), false)) {
    private val currencyKey = '¤'

    private val separators = listOf('.', ',')

    private var currencyFormatter: NumberFormat = NumberFormat
        .getCurrencyInstance(locale)
        .clearFormatter()

    private val decimalFormatSymbols: DecimalFormatSymbols
        get() = (currencyFormatter as DecimalFormat).decimalFormatSymbols

    private val String.withNbsp: String
        get() = replace(' ', ' ')

    fun updateCurrencyParams(locale: Locale) {
        val currentValue = rawValue.replace(decimalFormatSymbols.decimalSeparator, '.')

        currencyFormatter = NumberFormat.getCurrencyInstance(locale).clearFormatter()

        val newValue = currentValue.replace('.', decimalFormatSymbols.decimalSeparator)

        applyChangeFrom(newValue)
    }

    private fun invalidateMaskDataForFormatted(forValue: Number) {
        val formatted = currencyFormatter.format(forValue)
        val maskPattern = formatPattern(formatted)

        val decoding = listOf(
            MaskKey(
                key = '#',
                filter = "\\d",
                placeholder = '0'
            ),
            MaskKey(
                key = decimalFormatSymbols.decimalSeparator,
                filter = "[${decimalFormatSymbols.decimalSeparator}]",
                placeholder = decimalFormatSymbols.decimalSeparator
            )
        )

        updateMaskData(MaskData(maskPattern, decoding, maskData.alwaysVisible), false)
    }

    override fun overrideRawValue(newRawValue: String) {
        val parsed = currencyFormatter.parse(newRawValue) ?: 0

        invalidateMaskDataForFormatted(parsed)

        super.overrideRawValue(newRawValue)
    }

    override fun applyChangeFrom(newValue: String, position: Int?) {
        val diff = TextDiff.build(value, newValue.withNbsp)

        val decimalSeparator = decimalFormatSymbols.decimalSeparator
        val oldSeparatorIndex = value.indexOfLast { it == decimalSeparator }
        val newSeparatorIndex = newValue.indexOfLast { it == decimalSeparator }
        val needInvalidateMask = oldSeparatorIndex != newSeparatorIndex ||
                (oldSeparatorIndex == -1 && newSeparatorIndex == -1)

        val clearedValue = newValue.toValidFormat(diff)

        val numberValue = currencyFormatter.parse(
            when {
                clearedValue.startsWith(decimalSeparator) ->
                    decimalFormatSymbols.zeroDigit + clearedValue
                clearedValue.isBlank() -> decimalFormatSymbols.zeroDigit.toString()
                else -> clearedValue
            }
        ) ?: 0

        cleanup(diff)

        if (needInvalidateMask) {
            invalidateMaskDataForFormatted(numberValue)
        }

        replaceChars(clearedValue, 0)

        cursorPosition =
            if (value.length > diff.start && value[diff.start] == decimalFormatSymbols.groupingSeparator) {
                position ?: cursorPosition
            } else {
                abs(value.length - (newValue.length - (position ?: cursorPosition)))
            }.coerceAtMost(value.length)
    }

    private fun NumberFormat.clearFormatter(): NumberFormat {
        (this as? DecimalFormat)?.apply {
            val pattern = toPattern().filter { it != currencyKey }.trim()
            applyPattern(pattern)
        }

        return this
    }

    private fun String.toValidFormat(diff: TextDiff): String {
        val separatorChar = decimalFormatSymbols.decimalSeparator

        var separatorOutOfDiffIndex: Int = -1

        var index = 0

        while (index < this.length) {
            if (this[index] == separatorChar && !inDiff(diff, index)) {
                separatorOutOfDiffIndex = index

                break
            }

            index++
        }

        val replaceCharInDiff =
            if (diff.added == 1 && diff.removed == 0) {
                val diffChar = this[diff.start]

                if (diffChar in separators) diff.start else -1
            } else -1

        val maxSeparatorOffset = currencyFormatter.maximumFractionDigits

        var leftToInsert = maxSeparatorOffset

        if (separatorOutOfDiffIndex != -1) {
            index = separatorOutOfDiffIndex

            while (index < length) {
                if (this[index].isDigit() && !inDiff(diff, index)) leftToInsert--
                index++
            }
        } else {
            var oldSeparatorLeft = false

            forEachIndexed { i, char ->
                val inDiff = inDiff(diff, i)

                when {
                    char == separatorChar -> {
                        oldSeparatorLeft = true
                    }
                    !inDiff && oldSeparatorLeft && char.isDigit() -> {
                        leftToInsert--
                    }
                }
            }
        }

        val containsSeparator = contains(separatorChar) || replaceCharInDiff != -1

        return buildString {
            index = this@toValidFormat.length - 1

            var separatorInserted = false

            while (index >= 0) {
                val char = this@toValidFormat[index]

                val canInsertSeparator = length <= maxSeparatorOffset

                if (char.isDigit()) {
                    if (inDiff(diff, index) && !separatorInserted && containsSeparator) {
                        if (leftToInsert > 0) {
                            append(char)

                            leftToInsert--
                        }
                    } else {
                        append(char)
                    }
                } else if (canInsertSeparator && separatorOutOfDiffIndex == -1 && index == replaceCharInDiff) {
                    append(separatorChar)

                    separatorInserted = true
                } else if (canInsertSeparator && char == separatorChar && (separatorOutOfDiffIndex == index || separatorOutOfDiffIndex == -1)) {
                    append(separatorChar)

                    separatorInserted = true
                    separatorOutOfDiffIndex = index
                }

                index--
            }
        }
            .reversed()
            .let {
                index = it.indexOfFirst { char -> char != decimalFormatSymbols.zeroDigit }

                if (index <= 0) return@let it

                val firstNotZeroChar = it[index]

                if (firstNotZeroChar == decimalFormatSymbols.decimalSeparator) {
                    it.drop(index - 1)
                } else {
                    it.drop(index)
                }
            }
    }

    private fun inDiff(diff: TextDiff, index: Int): Boolean =
        diff.start <= index && index < diff.start + diff.added

    private fun formatPattern(pattern: String): String {
        return buildString {
            pattern.forEach { char ->
                when {
                    char.isDigit() -> append('#')
                    else -> append(char)
                }
            }
        }
    }

    override fun onException(exception: Exception) {
        onError.invoke(exception)
    }
}

File: client/web/divkit/src/utils/mask/currencyInputMask.ts
import type { LogError, WrappedError } from '../wrapError';
import { BaseInputMask } from './baseInputMask';
import { TextDiff, textDiff } from '../textDiff';
import { wrapError } from '../wrapError';

export class CurrencyInputMask extends BaseInputMask {
    protected currencyFormatter = new Intl.NumberFormat();
    protected decimalSeparator = '.';
    protected localeDigits: Record<string, string> = {};
    protected trimZeroRegExp = new RegExp('');

    constructor(locale: string | undefined = undefined, private readonly logError: LogError) {
        super({
            pattern: '',
            decoding: [],
            alwaysVisible: false
        });
        this.initFormatter(locale);
    }

    updateCurrencyParams(locale?: string) {
        const currentValue = this.parseFormat(this.rawValue) || 0;
        this.initFormatter(locale);

        const newValue = currentValue.toString().replace('.', this.decimalSeparator);

        this.applyChangeFrom(newValue);
    }

    protected initFormatter(locale?: string): void {
        try {
            this.currencyFormatter = new Intl.NumberFormat(locale, {
                minimumFractionDigits: 2,
                maximumFractionDigits: 2
            });
            this.decimalSeparator = this.currencyFormatter.format(0)[1];
            this.localeDigits = new Array(10).fill('').reduce((acc, _, i) => {
                acc[i] = this.currencyFormatter.format(i)[0];
                return acc;
            }, {});
            this.trimZeroRegExp = new RegExp(`^${this.localeDigits['0']}+`);
        } catch (err) {
            this.onException(wrapError(err as Error, {
                level: 'error',
                additional: {
                    locale
                }
            }));
        }
    }

    protected invalidateMaskDataForFormatted(forValue: number): void {
        const formatted = this.currencyFormatter.format(forValue);
        const pattern = this.formatPattern(formatted);
        const decoding = [{
            key: '#',
            filter: `[${[...Object.values(this.localeDigits)].join('')}]`,
            placeholder: this.localeDigits[0]
        }, {
            key: this.decimalSeparator,
            filter: `[${this.decimalSeparator}]`,
            placeholder: this.decimalSeparator
        }];

        this.updateMaskData({
            pattern,
            decoding,
            alwaysVisible: this.maskData.alwaysVisible
        }, false);
    }

    overrideRawValue(newRawValue: string): void {
        const parsed = this.parseFormat(newRawValue) || 0;

        this.invalidateMaskDataForFormatted(parsed);

        super.overrideRawValue(newRawValue);
    }

    applyChangeFrom(newValue: string, position?: number): void {
        const diff = textDiff(this.value, newValue);
        const oldSeparatorIndex = this.value.lastIndexOf(this.decimalSeparator);
        const newSeparatorIndex = newValue.lastIndexOf(this.decimalSeparator);
        const needInvalidateMask =
            oldSeparatorIndex !== newSeparatorIndex ||
            (oldSeparatorIndex === -1 && newSeparatorIndex === -1);
        const clearedValue = this.validFormat(newValue, diff);

        this.cleanup(diff);

        const rawValue = this.parseFormat(clearedValue) || 0;

        if (needInvalidateMask) {
            this.invalidateMaskDataForFormatted(rawValue);
        }

        this.replaceChars(clearedValue, 0);

        if (this.value.length > diff.start && !this.isDigit(this.value[diff.start])) {
            this.cursorPos = position ?? this.cursorPosition;
        } else {
            this.cursorPos = Math.abs(this.value.length - (newValue.length - (position ?? this.cursorPosition)));
        }
    }

    protected parseFormat(value: string): number {
        return parseFloat(
            value.replace(/./g, char => {
                const localeDigit = this.localeDigits[char];

                if (localeDigit) {
                    return localeDigit;
                }

                if (char === this.decimalSeparator) {
                    return '.';
                }

                return '';
            })
        );
    }

    protected formatPattern(pattern: string): string {
        let result = '';

        for (const char of pattern) {
            result += this.isDigit(char) ? '#' : char;
        }

        return result;
    }

    protected validFormat(value: string, diff: TextDiff): string {
        if (!value) {
            return '';
        }

        let separatorOutOfDiffIndex = -1;
        let index = 0;

        while (index < value.length) {
            if (value[index] === this.decimalSeparator && !this.inDiff(diff, index)) {
                separatorOutOfDiffIndex = index;
                break;
            }

            index++;
        }

        let replaceCharInDiff = -1;

        if (diff.added === 1 && diff.removed === 0 && [',', '.'].includes(value[diff.start])) {
            replaceCharInDiff = diff.start;
        }

        const maxSeparatorOffset = this.currencyFormatter.resolvedOptions().maximumFractionDigits;
        let leftToInsert = maxSeparatorOffset;

        if (separatorOutOfDiffIndex !== -1) {
            index = separatorOutOfDiffIndex;

            while (index < value.length) {
                if (this.isDigit(value[index]) && !this.inDiff(diff, index)) {
                    leftToInsert--;
                }
                index++;
            }
        } else {
            let oldSeparatorLeft = false;

            for (let i = 0; i < value.length; i++) {
                const char = value[i];
                if (char === this.decimalSeparator) {
                    oldSeparatorLeft = true;
                } else if (!this.inDiff(diff, i) && oldSeparatorLeft && this.isDigit(char)) {
                    leftToInsert--;
                }
            }
        }

        const containsSeparator = value.includes(this.decimalSeparator) || replaceCharInDiff !== -1;
        const result: string[] = [];

        index = value.length - 1;
        let separatorInserted = false;

        while (index >= 0) {
            const char = value[index];
            const canInsertSeparator = result.length <= maxSeparatorOffset;

            if (this.isDigit(char)) {
                if (this.inDiff(diff, index) && !separatorInserted && containsSeparator) {
                    if (leftToInsert > 0) {
                        result.push(char);
                        leftToInsert--;
                    }
                } else {
                    result.push(char);
                }
            } else if (canInsertSeparator && separatorOutOfDiffIndex === -1 && index === replaceCharInDiff) {
                result.push(this.decimalSeparator);
                separatorInserted = true;
            } else if (
                canInsertSeparator &&
                char === this.decimalSeparator &&
                (separatorOutOfDiffIndex === index || separatorOutOfDiffIndex === -1)
            ) {
                result.push(this.decimalSeparator);

                separatorInserted = true;
                separatorOutOfDiffIndex = index;
            }

            index--;
        }

        return result.reverse().join('').replace(this.trimZeroRegExp, '');
    }

    protected inDiff(diff: TextDiff, index: number): boolean {
        return diff.start <= index && index < diff.start + diff.added;
    }

    protected isDigit(char: string): boolean {
        return Boolean(this.localeDigits[char]);
    }

    onException(exception: WrappedError) {
        this.logError(exception);
    }
}
